scripts/check-components

```diff
@@ -0,0 +1,258 @@
+#!/usr/bin/ruby
+# coding: utf-8
+
+require 'find'
+
+COMMON_EVENTS = ['EventsReceived', 'EventsSent']
+SOURCE_EVENTS = ['BytesReceived'] + COMMON_EVENTS
+SINK_EVENTS = COMMON_EVENTS + ['BytesSent']
+# Transforms implicitly handle the common events in the topology, and
+# have no other emit requirements.
+TRANSFORM_EVENTS = []
+
+$error_count = 0
+
+def hash_array_add(hash, key, item)
+  arr = hash.fetch(key, Array::new)
+  arr.append(item)
+  hash[key] = arr
+end
+
+# Reduce a path to a component name, used to group all the emits for a
+# component that may be split over multiple files.
+def path_component(path)
+  parts = path.split('/')
+  parts.shift # Drop the leading 'src'
+  group = parts.shift
+  component = parts.shift
+
+  # Skip the top-level module files and utilities
+  if component == 'mod.rs' or component == 'util'
+    return nil
+  end
+
+  if component == 'gcp'
+    component = "gcp_#{parts.shift}"
+  end
+
+  if component.end_with? '.rs'
+    component = component[..-4]
+  end
+  return "#{group}/#{component}"
+end
+
+# A class to hold error reports and common functionality
+class Event
+  def initialize(name)
+    @path = nil
+    @name = name
+    @reports = []
+    @members = []
+    @counters = []
+    @logs = []
+  end
+
+  def path(path)
+    @path = path
+  end
+
+  def members(list)
```

Comment:
```suggestion
  def members=(list)
```

This can also be achieved with a class-level declaration of:

```ruby
class Event
  attr_writer :events
```

`attr_reader` and `attr_accessor` are also available.

Reply:
Got it.